project-koku/koku
# Cost 2487 fix db perf sorting

Fixes [COST-2487](https://issues.redhat.com/browse/COST-2487)

Changes proposed in this PR:
* Set connection to read-only
* Fix sorting to prioritize active connections and primary app database
* Add explain plan page

Testing Instructions:
1. Checkout this branch 
2. Stop koku `docker-compose stop`
3. Start postgres `docker-compose start db`
4. start `psql` and then execute the following
    ```sql
    \c template1
    create database koku template postgres;
    \c koku
    ```
5. Do not exit psql
6. Start koku `docker-compose start`
7. Start a second `psql` session and execute
    ```sql
    select pg_sleep(120);
    ```
8. Back to first `psql`, execute
    ```sql
    begin;
    update api_provider set paused = false;
    ```
9. Navigate to `http://localhost:5042/api/cost-management/v1/db-performance/stat-activity/`
    You should see `postgres` database at the top with "idle in transaction" state followed by `koku` database in "active" state
10. Navigate to `http://localhost:5042/api/cost-management/v1/db-performance/explain-query/`
    You should see a page with an editable div at the top and a display div with a gray background at the bottom.
11. Enter a query and click Explain (please try to sneak in something dangerous here)
    1. If you entered a valid query, you should see a query plan
    2. If you entered an invalid query, you  should see a `ProgrammingError`
    3. If you entered multiple queries, you should see a query plan for each query
    4. If you entered a DDL statement (ALTER, CREATE, DROP) You should see a message that refuses to process a DDL statement
    5. If you entered COMMIT, ROLLBACK, You should see a refusal to process that statement
    6. If yo entered an INSERT UPDATE or DELETE You should also see a refusal to process the statement
12. You can highlight a single statement in a series of statements and click Explain to get a query plan 
13. You can press `Ctrl+Enter` instead of clicking Explain
14. Clear the query input and Explain to see "Nothing to process"

---
Additional Context:
